File: CSharp/Eta/Src/ValueAdd/Rdm/MsgBase.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2022-2023 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

using LSEG.Eta.Codec;
using System.Text;

namespace LSEG.Eta.ValueAdd.Rdm
{
    /// <summary>
    /// Base class for RDM message classes.
    /// </summary>
    public abstract class MsgBase
    {
        private StringBuilder stringBuf = new StringBuilder();

        /// <summary>
        /// Defines end of line character.
        /// </summary>
        protected string eol = System.Environment.NewLine;

        /// <summary>
        /// Define tabulation character.
        /// </summary>
        protected const string tab = "\t";

        /// <summary>
        /// The ID of the stream the message is sent to.
        /// </summary>
        public virtual int StreamId { get; set; }
        /// <summary>
        /// The domain type of the message.
        /// </summary>
        public virtual int DomainType { get; set; }
        /// <summary>
        /// Message Class <see cref="MsgClasses"/>
        /// </summary>
        public virtual int MsgClass { get; }

        /// <summary>
        /// Clears current message instance.
        /// </summary>
        public virtual void Clear()
        {
            StreamId = 0;
        }

        /// <summary>
        /// Encode an RDM message.
        /// </summary>
        /// <param name="encodeIter"><see cref="EncodeIterator"/> instance that performs the encoding.</param>
        /// <returns><see cref="CodecReturnCode"/> value indicating the status of the operation.</returns>
        public abstract CodecReturnCode Encode(EncodeIterator encodeIter);

        /// <summary>
        /// Decode a ETA message into an RDM message.
        /// </summary>
        /// <param name="decodeIter"><see cref="DecodeIterator"/> instance that performs the decoding.</param>
        /// <param name="msg">Partially decoded message.</param>
        /// <returns><see cref="CodecReturnCode"/> indicating the status of the operation</returns>
        public abstract CodecReturnCode Decode(DecodeIterator decodeIter, Msg msg);

        /// <summary>
        /// Internal virtual function to initialize the toString functionality.
        /// </summary>
        /// <returns>The <see cref="StringBuilder"/> object</returns>
        protected virtual StringBuilder PrepareStringBuilder()
        {
            stringBuf.Clear();

            stringBuf.Append(tab);
            stringBuf.Append("streamId: ");
            stringBuf.Append(StreamId);
            stringBuf.Append(eol);

            return stringBuf;
        }
    }
}

File: Java/Eta/ValueAdd/src/main/java/com/refinitiv/eta/valueadd/domainrep/rdm/MsgBase.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2019-2022 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.eta.valueadd.domainrep.rdm;

import com.refinitiv.eta.codec.DecodeIterator;
import com.refinitiv.eta.codec.EncodeIterator;
import com.refinitiv.eta.codec.Msg;

/**
 * This message structure contains the information that is common across all RDM
 * Message formats. It is included in all Value Added RDM Components.
 */
public interface MsgBase
{
    /**
     * The Stream Id for the given item.
     * 
     * @return the streamId
     */
    public int streamId();

    /**
     * The Stream Id for the given item.
     *
     * @param streamId the streamId to set
     * 
     */
    public void streamId(int streamId);
    
    /**
     * Returns the domain type of the RDM message.
     *
     * @return the int
     */
    public int domainType();

    /**
     * Encode an RDM message.
     * 
     * @param eIter The Encode Iterator
     * 
     * @return ETA return value
     */
    public int encode(EncodeIterator eIter);

    /**
     * Decode a ETA message into an RDM message.
     *
     * @param dIter The Decode Iterator
     * @param msg the msg
     * @return ETA return value
     */
    public int decode(DecodeIterator dIter, Msg msg);

    /**
     * Clears the current contents of the message and prepares it for re-use.
     */
    public void clear();
}